### PR TITLE
ARROW-247: Missing explicit destructor in RowBatchReader causes an incomplete type error

### PR DESCRIPTION
**Pull request by jihoonson**

### CODE REVIEW

**Comment by jihoonson** [comment]
I have no idea why travis is not triggered. Is there any problem?

**Comment by wesm** [comment]
Appears to not be triggering for other PRs too. Seems that Travis has been having some infra problems

**Comment by wesm** [comment]
lgtm otherwise

**Comment by jihoonson** [comment]
Thanks. I added one more commit to trigger travis, but it doesn't respond. It seems a travis' infra problem as you said.

**Comment by jihoonson** [comment]
I don't know why travis fails.. My first commit passes all tests in my local travis (https://travis-ci.org/jihoonson/arrow/builds/149333166), but after adding some comments, travis continuously fails.

File: cpp/src/arrow/ipc/adapter.h [hunk 1]
```diff
@@ -78,6 +78,8 @@ class ARROW_EXPORT RowBatchReader {
   static Status Open(MemorySource* source, int64_t position, int max_recursion_depth,
       std::shared_ptr<RowBatchReader>* out);
 
+  ~RowBatchReader();
```
Comment (xhochy) [inline cpp/src/arrow/ipc/adapter.h]: Please make the constructor virtual. This is probably not the problem why Travis fails but at least a problem.

**Comment by wesm** [comment]
The failure is being caused by https://github.com/apache/parquet-cpp/commit/38f0ffd5adc37db948991a2eb6409e2727721463

@edani any ideas off the top of your head? This should be reproducible

**Comment by wesm** [comment]
Google seems to suggest this is an alignment issue

**Comment by wesm** [comment]
It seems that these bit-packing subroutines will fail if the read batch size is not a multiple of 32. I will have to look more closely later today

**Comment by wesm** [comment]
Strike that, not sure what's going on (if the read size is < 32 in `unpack32` then it should not attempt to do any unpacking). I'll try to debug in gdb if no one beats me to it

**Comment by edani** [comment]
I can't seem to reproduce the problem, running this:

```
valgrind --tool=memcheck --leak-check=full ./build/debug/parquet-reader-writer-test 
```

It could be that I didn't build arrow the same way as travis did...

**Comment by wesm** [comment]
@edani it appears that arrow and parquet are being built with different architecture flags. Arrow is being built with `-msse3` and parquet-cpp without any such flags. Could this possibly cause the conflict?

**Comment by wesm** [comment]
I'm able to reproduce the core dump locally from the .so file that Arrow is using to run its tests. I will investigate

**Comment by wesm** [comment]
I pushed a new build via PARQUET-672. I wasn't able to reproduce the error with a complete local build of both parquet-cpp and the arrow libraries. May have to roll back PARQUET-671 until a more thorough investigation can take place

**Comment by wesm** [comment]
@jihoonson could you rebase and force push to get a new build? I'm chalking this up to transient badness in the Travis CI environment, not sure what else it could be.

File: cpp/src/arrow/ipc/adapter.h (hunk 1 above)
Comment (jihoonson) [inline cpp/src/arrow/ipc/adapter.h]: Thanks. I made it virtual.

**Comment by jihoonson** [comment]
@wesm and @edani, appreciate your investigation. Travis now seems ok and all tests passed successfully.

**Comment by wesm** [comment]
+1

**Pull request closed without merging**